### DISCUSSION TITLE
what if a rule action failed

### DISCUSSION THREAD

**Discussion by robodevchampion** [lf-edge/ekuiper#3057] [Q&A] [answered]
**Please describe your problem in detail, if necessary, you can upload the log file through the attachment**:

Hi, Lets say, I have a rule that process incoming stream of messages, it has some actions, lets say for Redis or MQTT. Now due to connectivity loss (offline) it failed to push result. 

So those messages are lost ? or is there any way to recover them ?

Looking for inbuilt support for offline support. 

Thanks
Abhishek

**Comment by ngjaying** [comment]
Please check the sink cache https://ekuiper.org/docs/en/latest/guide/sinks/overview.html#caching

**Comment by robodevchampion** [comment, marked as answer]
I tried to use cache, so what I do is I have 2 broker one is for source & other for sink. 

first I test rule it is working as expected, 

then I stop sink broker for some time, and push some messages on source broker, rule shows me error -> MQTT connection lost, 

now i restart sink broker, so expectation is to get messages which were sent when this broker is offline, but nothing happened.

Is it right way of testing ? or am I missing something 

Thanks
Abhishek Bhatia

**Reply by ngjaying** [reply]
Push some message in source side after some time. The MQTT reconnection takes time and the resend is triggered by new message.

**Reply by robodevchampion** [reply]
It worked, Thanks
